freetextresponse/freetextresponse.py

```diff
@@ -0,0 +1,437 @@
+            paths_js=[],
+            urls_css=[],
+            urls_js=[],
+            fragment_js=None,
+    ):
+        #  pylint: disable=dangerous-default-value, too-many-arguments
+        '''
+        Assemble the HTML, JS, and CSS for an XBlock fragment
+        '''
+        fragment = Fragment(html_source)
+        for url in urls_css:
+            fragment.add_css_url(url)
+        for path in paths_css:
+            url = self.get_resource_url(path)
+            fragment.add_css_url(url)
+        for url in urls_js:
+            fragment.add_javascript_url(url)
+        for path in paths_js:
+            url = self.get_resource_url(path)
+            fragment.add_javascript_url(url)
+        if fragment_js:
+            fragment.initialize_js(fragment_js)
+        return fragment
+
+    def get_indicator_visiblity_class(self):
+        '''
+        Returns the visibility class for the correctness indicator html element
+        '''
+        if re.search('false', self.display_correctness, re.IGNORECASE):
+            return 'hidden'
+        return ''
+
+    def get_word_count_message(self):
+        '''
+        Returns the word count message based on the student's answer
+        '''
+        if self.num_attempts < 1 or self.word_count_valid():
+            return ''
+        else:
+            return "Invalid Word Count. Your reponse must be " \
+                   "between {min} and {max} words."\
+                .format(
+                    min=self.min_word_count,
+                    max=self.max_word_count,
+                )
+
+    def get_indicator_class(self):
+        '''
+        Returns the class of the correctness indicator element
+        '''
+        if self.student_answer == '':
+            if self.num_attempts == 0:
+                return 'unanswered'
+            else:
+                return 'incorrect'
+        elif (
+                self.word_count_valid() and
+                (
+                    self.phrase_present_in_answer(
+                        self.halfcredit_keyphrases
+                    )
+                    or
+                    self.phrase_present_in_answer(
+                        self.fullcredit_keyphrases
+                    )
+                )
+        ):
+            return 'correct'
+        else:
+            return 'incorrect'
+
+    def word_count_valid(self):
+        '''
+        Returns a boolean value indicating whether the current
+        word count of the user's answer is valid
+        '''
+        word_count = len(self.student_answer.split())
+        return (
+            word_count <= self.max_word_count and
+            word_count >= self.min_word_count
+        )
+
+    def phrase_present_in_answer(self, phrase_list):
+        '''
+        Helper method that returns a boolean value indicating
+        whether at least one member of the provided phrase list
+        is present in the student's answer
+        '''
+        if re.search(
+                '|'.join(
```

Comment:
Dynamically generated regexs; eek!

In this context, I'm not sure regex's are the best tool for the task.
They're not super efficient and this could be a pain to debug;
what happens if my phrase contains a pipe `|` or some other special chars??

A more pythonic approach would be something like

```python
def is_at_least_one_phrase_present(phrases, answer):
    answer = answer.lower()
    matches = [
        phrase.lower() in answer
        for phrase in phrases
    ]
    return any(matches)
```